### PR TITLE
Dont die when we're missing a route

### LINKED ISSUE #19145: Upgrade to 8.1.3 fails with URL route error
Trying to update to 8.1.3, I get the following error, and the update fails:

```
Symfony\Component\Routing\Exception\RouteNotFoundException: Unable to generate a URL for the named route "calendar_index" as such route does not exist.
```

Here are the last few lines of owncloud.log, which seem related:

```
"reqId":"xVav1QqMQl3YXheDBhin","remoteAddr":"136.187.82.195","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 503 Service unavailable\",\"Code\":0,\"Trace\":\"#0 [internal function]: {closure}(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\\n#2 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/var\\\/www\\\/owncloud\\\/remote.php(64): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/var\\\/www\\\/owncloud\\\/remote.php(135): handleException(Object(RemoteException))\\n#6 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/remote.php\",\"Line\":55}","level":4,"time":"2015-09-18T05:29:27+00:00"}
{"reqId":"HHQHYqifIkZjol5gf6SL","remoteAddr":"136.187.82.195","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 503 Service unavailable\",\"Code\":0,\"Trace\":\"#0 [internal function]: {closure}(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\\n#2 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/var\\\/www\\\/owncloud\\\/remote.php(64): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/var\\\/www\\\/owncloud\\\/remote.php(135): handleException(Object(RemoteException))\\n#6 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/remote.php\",\"Line\":55}","level":4,"time":"2015-09-18T05:29:59+00:00"}
{"reqId":"FmiElc+g26BN8UVAcijg","remoteAddr":"136.187.82.195","app":"index","message":"Exception: {\"Exception\":\"OC\\\\NeedsUpdateException\",\"Message\":\"\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/app.php(98): OC_App::loadApp('activity')\\n#1 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/util.php(127): OC_App::loadApps(Array)\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/base.php(873): OC_Util::setupFS()\\n#3 \\\/var\\\/www\\\/owncloud\\\/index.php(40): OC::handleRequest()\\n#4 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/app.php\",\"Line\":118}","level":4,"time":"2015-09-18T05:30:10+00:00"}
{"reqId":"1TMm\/QK7ffnXVTTvxCVm","remoteAddr":"136.187.82.195","app":"update","message":"Exception: {\"Exception\":\"Symfony\\\\Component\\\\Routing\\\\Exception\\\\RouteNotFoundException\",\"Message\":\"Unable to generate a URL for the named route \\\"calendar_index\\\" as such route does not exist.\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(306): Symfony\\\\Component\\\\Routing\\\\Generator\\\\UrlGenerator->generate('calendar_index', Array, false)\\n#1 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/urlgenerator.php(65): OC\\\\Route\\\\Router->generate('calendar_index', Array)\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/helper.php(139): OC\\\\URLGenerator->linkToRoute('calendar_index', Array)\\n#3 \\\/var\\\/www\\\/owncloud\\\/lib\\\/public\\\/util.php(306): OC_Helper::linkToRoute('calendar_index', Array)\\n#4 \\\/var\\\/www\\\/owncloud\\\/apps\\\/calendar\\\/appinfo\\\/app.php(30): OCP\\\\Util::linkToRoute('calendar_index')\\n#5 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/app.php(139): require_once('\\\/var\\\/www\\\/ownclo...')\\n#6 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/app.php(120): OC_App::requireAppFile('calendar')\\n#7 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/app.php(1173): OC_App::loadApp('calendar', false)\\n#8 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/installer.php(218): OC_App::updateApp('calendar')\\n#9 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/installer.php(248): OC_Installer::updateApp(Array)\\n#10 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/updater.php(487): OC_Installer::updateAppByOCSId('168707')\\n#11 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/updater.php(308): OC\\\\Updater->upgradeAppStoreApps(Array)\\n#12 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/updater.php(195): OC\\\\Updater->doUpgrade('8.1.3.0', '8.1.1.3')\\n#13 \\\/var\\\/www\\\/owncloud\\\/core\\\/ajax\\\/update.php(87): OC\\\\Updater->upgrade()\\n#14 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/symfony\\\/routing\\\/Symfony\\\/Component\\\/Routing\\\/Generator\\\/UrlGenerator.php\",\"Line\":134}","level":4,"time":"2015-09-18T05:30:26+00:00"}
```

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #19145

Better to generate one invalid url then die

cc @PVince81 @Xenopathic

### CODE REVIEW

**Comment by nickvergessen** [comment]
Fine by me :+1:

File: lib/private/route/router.php [hunk 1]
```diff
@@ -311,11 +321,17 @@ public function getGenerator() {
 	 */
 	public function generate($name, $parameters = array(), $absolute = false) {
 		$this->loadRoutes();
-		return $this->getGenerator()->generate($name, $parameters, $absolute);
+		try {
+			return $this->getGenerator()->generate($name, $parameters, $absolute);
+		} catch (RouteNotFoundException $e) {
+			$this->logger->error($e->getMessage());
```
Comment (MorrisJobke) [inline lib/private/route/router.php]: Can we log this as exception? Use `logException()`

Comment (icewind1991) [inline lib/private/route/router.php]: Forgot about that method, fixed

**Comment by MorrisJobke** [comment]
@icewind1991 Please check out the failing unit tests

**Comment by DeepDiver1975** [comment]
> @icewind1991 Please check out the failing unit tests

```
OC\AppFramework\Routing\RoutingTest::testSimpleRoute
Argument 1 passed to OC\Route\Router::__construct() must implement interface OCP\ILogger, none given, called in phar:///usr/local/bin/phpunit/phpunit-mock-objects/Framework/MockObject/Generator.php on line 254 and defined

/ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/lib/private/route/router.php:88
/ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/tests/lib/appframework/routing/RoutingTest.php:127
/ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/tests/lib/appframework/routing/RoutingTest.php:18

**Comment by PVince81** [comment]
Tested passed, :+1:

**Comment by MorrisJobke** [comment]
@karlitschek @icewind1991 Should we backport this, because the initial issue was put into the 8.1.5 milestone https://github.com/owncloud/core/issues/19145

**Comment by karlitschek** [comment]
yes. olease backport please

**Comment by DeepDiver1975** [comment]
> yes. olease backport please

@icewind1991 please open backport PR

**Comment by icewind1991** [comment]
- 8.2: #20830
- 8.1: #20831

**Pull request merged**